### DISCUSSION TITLE
I would like a similar method in this package, and I don't know if it exists.

### DISCUSSION THREAD

**Discussion by osquelsantisteban** [spatie/laravel-permission#2138] [Q&A]
I would like a similar method in this package, and I don't know if it exists. Where it receives a search parameter if it exists and is like a like and not a where.

```
// $this->buscar   text to Search

//! get_roles
    public function get_roles(){

        $roles = Role::all();
        $rolSearch = [];
        if($this->buscar){
            foreach($roles as $rol){
                if(Str::contains(Str::upper($rol->getAttribute('name')), Str::upper($this->buscar)))
                {
                    $rolSearch[] = $rol;
                }
            }
        }
        else
            return $roles;

        return $rolSearch;
    }
```

**Comment by fredsal** [comment]
mamma mia, spaghetti 
that's the worst code i've seen in some time

use early return, use collection native methods

**Reply by osquelsantisteban** [reply]
that resolve my problem immediate, if you have a solution better show me please

**Reply by fredsal** [reply]
> use early return, use collection native methods

And use google if you don't know how to use these best practices
https://medium.com/swlh/return-early-pattern-3d18a41bba8
https://laravel.com/docs/9.x/collections#available-methods

Also use markdown correctly for highlighting ` ```php `

**Reply by systemsolutionweb** [reply]
You can't force people to smart code if they don't know best practices and the tools they have at their disposal, also if they don't have the capacity.

Just use `like`, also why is `buscar` a propertie? it is better if it is just an argument `get_roles($buscar = null)`
```php
public function get_roles($buscar = null){
    return Role::where('name', 'like', '%' . ((string) $buscar) . '%')->get();
}
```
I don't want to imagine how the rest of the code is 😄

**Reply by osquelsantisteban** [reply]
that not resolve the problem of the laravel-permission, only you are focusing on my code instead of solving the problem or getting the idea of ​​what I want to do, that my doubt is not only mine.

**Reply by systemsolutionweb** [reply]
It is not a laravel-permission problem, only you are doing a big piece of code instead of a laravel single line of code
```php
Role::where('name', 'like', '%' . ((string) $buscar) . '%')->get();
```
Please don't reinvent the wheel, just use laravel
Also, duplicate of #2133(I assume it)

**Reply by osquelsantisteban** [reply]
that's exactly what I need but that laravel-permissions already have it built in. For when I have a table of users I need to search by any column eg name, email and role etc...

**Reply by systemsolutionweb** [reply]
Laravel and this package have all the methods, you only have to read documentation and use them
https://laravel.com/docs/9.x/eloquent
https://laravel.com/docs/9.x/eloquent-relationships
https://spatie.be/docs/laravel-permission/v5/basic-usage/basic-usage

**Reply by osquelsantisteban** [reply]
This not work 
```php
    public function get_usuarios(){

        $buscar = $this->buscar;
        $usuarios = User::buscar($this->buscar)
            ->with('ueb','roles')
            ->whereHas('roles',function($query) use ($buscar){
                $query->where('name', 'like', '%' . ((string) $buscar) . '%');
            })
            ->orderBy($this->sort,$this->direction)
            ->paginate($this->cantidad);
        return $usuarios;
    }

**Reply by systemsolutionweb** [reply]
Sure, let everybody use our psychic powers to know what exception you get, and what the `User::buscar()` method does

**Reply by osquelsantisteban** [reply]
Sorry i forgot, scopeBuscar , but this work ok, this whereHas is the one that doesn't work

``
->whereHas('roles',function($query) use ($buscar){
                $query->where('name', 'like', '%' . ((string) $buscar) . '%');
            })
``

```php

/*  Model User   */
class User extends Authenticatable
{
    use HasApiTokens;
    use HasFactory;
    use HasProfilePhoto;
    use Notifiable;
    use TwoFactorAuthenticatable;
    use HasRoles;
     ...

    /* Scope Buscar */
    public function scopeBuscar( $query, $buscar){
        if($buscar == '') return;

        return $query
        ->orwhere('id', 'like', "%{$buscar}%")
        ->orwhere('email', 'like', "%{$buscar}%")
        ->orwhere('name', '=', "{$buscar}")
        ->orwhereIn('ueb_id', function($query) use ($buscar){
                $query->select('id')->from('uebs')->where('nombre','like',"%{$buscar}%")->get();
        })
        ;
    }



     /*  Controller  */

    //! get_usuarios
    public function get_usuarios(){

        $buscar = $this->buscar;
        $usuarios = User::buscar($this->buscar)
            ->with('ueb','roles')
            ->whereHas('roles',function($query) use ($buscar){
                $query->where('name', 'like', '%' . ((string) $buscar) . '%');
            })
            ->orderBy($this->sort,$this->direction)
            ->paginate($this->cantidad);
        return $usuarios;
    }

**Reply by fredsal** [reply]
You should also learn the basics of sql, you can't ask every two lines of code
Also, use google for the exception you get, up I'm sure you have ambiguity issues

**Reply by systemsolutionweb** [reply]
>  I'm sure you have ambiguity issues

+1

**Reply by osquelsantisteban** [reply]
I solved

```php
public function scopeBuscar( $query, $buscar){
        if($buscar == '') return;

        return $query
        ->orwhere('id', 'like', "%{$buscar}%")
        ->orwhere('email', 'like', "%{$buscar}%")
        ->orwhere('name', '=', "{$buscar}")
        ->orwhereIn('ueb_id', function($query) use ($buscar){
                $query->select('id')->from('uebs')->where('nombre','like',"%{$buscar}%")->get();
        })
        ->when($buscar, function($q,$buscar){
            //Filter Role
            $rolesSearch = Role::where('name', 'like', '%' . ((string) $buscar) . '%')
                ->get()
                ->pluck('name');

            //Filter User by Role
                return User::role($rolesSearch);
        })
        ;
    }

**Reply by erikn69** [reply]
If it is fixed, mark as answered
